File: Plantemon/app/src/main/java/com/g4ng/logic/Move.java
package com.g4ng.logic;

import android.util.Log;
import com.g4ng.model.Plant;
import com.g4ng.model.Player;

public class Move implements Action {
    private final String name;
    private final int attack;
    private final int defense;
    private final int accuracy;

    public Move(String name, int attack, int defense, int accuracy) {
        this.name = name;
        this.attack = attack;
        this.defense = defense;
        this.accuracy = accuracy;
    }

    @Override
    public String execute(Player performer, Player opponent, Action opponentAction) {
        Plant attackerPlant = performer.getCurrentPlant();
        Plant targetPlant = opponent.getCurrentPlant();

        if (Math.random() * 100 <= accuracy) {
            int opponentDefense = (opponentAction != null) ? opponentAction.getDefenseValue() : 0;
            
            // Damage logic to increase dynamism
            double variation = 0.7 + (Math.random() * 0.4);
            int damage = (int)((this.attack - (opponentDefense * variation)) * variation);
            
            // Fix: Ensure damage is never negative (which would heal the opponent)
            damage = Math.max(0, damage);
            
            targetPlant.takeDamage(damage);
            String result = performer.getUsername()+ "'s " + attackerPlant.getName() + " used " + name + " and dealt " + damage + " damage!";
            Log.d("BattleLogic", result);
            return result;
        } else {
            String result = performer.getUsername()+ "'s " + attackerPlant.getName() + " missed " + name + "!";
            Log.d("BattleLogic", result);
            return result;
        }
    }

    @Override
    public int getDefenseValue() {
        return this.defense;
    }

    public String getName() {
        return name;
    }

    public int getAttack() {
        return attack;
    }

    public int getDefense() {
        return defense;
    }

    public int getAccuracy() {
        return accuracy;
    }
}

File: plantemon-web/src/lib/domain/move.ts
import { random } from "./rng";
import type { Action } from "./types";
import type { Player } from "./player";
import { requireCurrentPlant } from "./require-plant";

/**
 * Port of logic/Move.java.
 *
 * By design, higher-attack moves carry lower accuracy.
 */
export class Move implements Action {
  readonly name: string;
  readonly attack: number;
  readonly defense: number;
  readonly accuracy: number;

  constructor(name: string, attack: number, defense: number, accuracy: number) {
    this.name = name;
    this.attack = attack;
    this.defense = defense;
    this.accuracy = accuracy;
  }

  execute(performer: Player, opponent: Player, opponentAction: Action | null): string {
    const attackerPlant = requireCurrentPlant(performer);
    const targetPlant = requireCurrentPlant(opponent);

    // Matches Java's `Math.random() * 100 <= accuracy`.
    if (random() * 100 <= this.accuracy) {
      const opponentDefense = opponentAction !== null ? opponentAction.getDefenseValue() : 0;

      // Damage variation, to keep turns from being deterministic.
      const variation = 0.7 + random() * 0.4;

      // Math.trunc reproduces Java's (int) cast, which truncates toward zero.
      // The clamp matters: without it a high opponent defense produced negative
      // damage, which healed the target.
      const damage = Math.max(0, Math.trunc((this.attack - opponentDefense * variation) * variation));

      targetPlant.takeDamage(damage);
      return `${performer.getUsername()}'s ${attackerPlant.getName()} used ${this.name} and dealt ${damage} damage!`;
    }

    return `${performer.getUsername()}'s ${attackerPlant.getName()} missed ${this.name}!`;
  }

  getDefenseValue(): number {
    return this.defense;
  }

  getName(): string {
    return this.name;
  }

  getAttack(): number {
    return this.attack;
  }

  getDefense(): number {
    return this.defense;
  }

  getAccuracy(): number {
    return this.accuracy;
  }
}
